app/models/manageiq/providers/openstack/cloud_manager/upload_image_workflow.rb

```diff
@@ -0,0 +1,76 @@
+class ManageIQ::Providers::Openstack::CloudManager::UploadImageWorkflow < Job
+  def self.create_job(ext_management_system_id, image_id, url)
+    options = {
+      :ext_management_system_id => ext_management_system_id,
+      :url                      => url,
+      :image_id                 => image_id,
+    }
+
+    super(name, options)
+  end
+
+  def pre_upload_image
+    queue_signal(:upload_image)
+  end
+
+  def upload_image
+    ext_management_system_id, image_id, url = options.values_at(:ext_management_system, :image_id, :url)
+
+    EmsCloud.find(ext_management_system_id).with_provider_connection(:service => 'Image') do |service|
+      started_on = Time.now.utc
+      update_attributes!(:started_on => started_on)
+
+      image = service.images.find_by_id(image_id)
+      service.handle_upload(image, url)
```

Comment:
Does this do the upload synchronously?  That's not going to work as this will likely timeout just like we were worried about without the state machine if it takes >10 mins.

The idea is that your `upload_image` state would start the upload, add some context key to the options hash, and queue the next state which check if the upload is finished.  If it is finished it signals finish and if not it requeues itself for e.g. 1 minute in the future.

This means that no queue method being handled by a worker is doing any long term blocking calls so the upload could theoretically take hours and we'd be just fine.

Reply:
In order for this to work you need to start the upload to run async in the background, queue the poll_upload method and return out of this method.